text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1087 @@
+`(const) Trait` bounds cannot be made conditional with `#[cfg]`. The only real useful reason to have this is to support newer Rust versions with a cfg, and allow older Rust versions to compile
+the traits, just without const support. This is surmountable with proc macros that either generate two versions or just generate a different version depending on the Rust version.
+Since it's only necessary for a transition period while a crate wants to support both pre-const-trait Rust and
+newer Rust versions, this doesn't seem too bad. With a MSRV bump the proc macro usage can be removed again.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## What is the impact of not doing this?
+
+We would require everything that wants a const-equivalent to have duplicated traits and not
+use `const` fn at all, but use associated consts instead. Similarly this would likely forbid
+invoking builtin operators. This same concern had been brought up for the `const fn` stabilization
+[7 years ago](https://github.com/rust-lang/rust/issues/24111#issuecomment-385046163).
+
+Basically what we can do is create
+
+```rust
+trait ConstDefault {
+    const DEFAULT: Self;
+}
+```
+
+and require users to use
+
+```rust
+const FOO: Vec<u8> = ConstDefault::DEFAULT;
+```
+
+instead of
+
+```rust
+const fn FOO: Vec<u8> = Default::default();
+```
+
+This duplication is what this RFC is suggesting to avoid.
+
+Since it has already been possible to do all of this on stable Rust for years, and no major
+crates have popped and gotten used widely, I assume that is either because
+
+* it's too much duplication, or
+* everyone was waiting for the work (that this RFC wants to stabilize) to finish, or
+* both.
+
+So while it is entirely possible that rejecting this RFC and deciding not to go down this route
+will lead to an ecosystem for const operations to be created, it would result in duplication and
+inconsistencies that we'd rather like to avoid.
+
+Such an ecosystem would also make `const fn` obsolete, as every `const fn` can in theory be represented
+as a trait, it would just be very different to use from normal rust code and not really allow nice abstractions to be built.
+
+```rust
+const fn add(a: u32, b: u32) -> u32 { a + b }
+
+struct Add<const A: u32, const B: u32>;
+
+impl<const A: u32, const B:u32> Add<A, B> {
+    const RESULT: u32 = A + B;
+}
+
+const FOO: u32 = add(5, 6);
+const BAR: u32 = Add<5, 6>::RESULT;
+```
+
+## use `const Trait` bounds for conditionally-const, invent new syntax for always-const
+
+It may seem tempting to use `const fn foo<T: const Trait>` to mean what in this RFC is `(const) Trait`, and then add new syntax for bounds that allow using trait methods in const blocks.
+
+Examples of possible always const syntax:
+
+* `=const Trait`
+* `const const Trait` (lol)
+* `const(always) Trait` (`pub` like)
+* `const<true> Trait` (effect generic like)
+* `const! Trait`
```

Comment:
Is there a strong reason not to do that? Because an always-const syntax seems more than tempting to me, it seems like the objectively correction solution.

`const(always)` in particular seems very expressive to me.

Reply:
The reason I ask is actually that I found this section while writing my own version of this RFC, which goes all-in on `const Trait` as the default bound, and `const(always) Trait` as the stronger version.

[Rendered.](https://github.com/PoignardAzur/rust-rfcs/blob/2310414784eb7c2398d81f071cb40bc575942b90/text/0000-const-trait-impls.md)